src/psyclone/tests/psyir/nodes/dynamic_omp_task_directive_test.py

```diff
@@ -0,0 +1,3915 @@
+
+
+def test_omp_task_directive_read_index_shared_type(fortran_reader):
+    ''' Test the code generation correctly generates an error if an array index
+    of a read array is a shared variable.'''
+    code = '''
+    subroutine my_subroutine()
+        type :: x
+          integer, dimension(320, 10) :: A
+        end type
+        type(x) :: AA
+        integer, dimension(321, 10) :: B
+        integer :: i
+        integer :: j
+        integer :: k
+        do i = 1, 320, 32
+            k = 4
+            k = -2
+            k = k + 3
+            do j = 1, 32
+                AA%A(i, j) = AA%A(i, k)
+                AA%A(i, j) = B(1+i, j) + k
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0].children[3].children[3]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(loops[0])
+    ptrans.apply(loops[0].parent.parent)
+    with pytest.raises(GenerationError) as excinfo:
+        tree.lower_to_language_level()
+    assert ("Shared variable access used as an array index inside an "
+            "OMPTaskDirective which is not supported. Variable name is "
+            "'Reference[name:'k']'" in str(excinfo.value))
+
+
+def test_omp_task_directive_33(fortran_reader):
+    ''' Test the code generation fails when attempting to access an array
+    when using an array element as an index.'''
+    code = '''
+    subroutine my_subroutine()
+        type :: x
+          integer, dimension(320, 10) :: A
+        end type
+        type(x) :: AA
+        integer, dimension(10, 10) :: B
+        integer :: i
+        integer :: j
+        do i = 1, 10
+            do j = 1, 10
+                B(i, j) = AA%A(B(1,2), j) + 1
+            end do
+        end do
+        do i = 1, 10
+            do j = 1, 10
+                AA%A(i, j) = 0
+            end do
+        end do
+    end subroutine
+    '''
+    tree = fortran_reader.psyir_from_source(code)
+    ptrans = OMPParallelTrans()
+    strans = OMPSingleTrans()
+    tdir = DynamicOMPTaskDirective()
+    loops = tree.walk(Loop, stop_type=Loop)
+    loop = loops[0]
+    parent = loop.parent
+    loop.detach()
+    tdir.children[0].addchild(loop)
+    parent.addchild(tdir, index=0)
+    strans.apply(parent.children)
+    ptrans.apply(parent.children)
+    with pytest.raises(GenerationError) as excinfo:
+        tree.lower_to_language_level()
+    assert ("'ArrayReference' object is not allowed to appear in an array "
+            "index expression inside an OMPTaskDirective."
+            in str(excinfo.value))
+
+
+def test_omp_task_directive_34(fortran_reader):
+    ''' Test the code generation correctly generates the correct code if
+    code contains a type within a type .'''
```

Comment:
This is actually testing for some error condition.

Reply:
Changed the comment, this is very similar to _32 but I think only being on the rhs of an assignment results in a technically different error being thrown.